rfcs/0180-broken-package-removal.md

```diff
@@ -0,0 +1,94 @@
+## Broken Packages
+
+Packages that are unconditionally marked as broken on all platforms are candidates for removal.
+If a package has dependent packages, those dependent packages will also be marked as indirectly broken.
+If the dependent packages cannot function without the broken package, they should also be removed.
+For example, if a library is broken and an application depends on it, the application will also be removed.
+
+## Unmaintained Packages
+
+We will add an "unmaintained" warning for all packages with an empty maintainer field that do not have any dependent packages.
+Ideally, we will have automation or semi-automation in place to create pull requests for this process (see future work).
+
+If a NixOS module depends on any removed package and is non-functional without it, the module will be removed as well.
+In `pkgs/top-level/aliases.nix`, we can link to the pull request that removed the package, making it easier for users to recover the nix expression if needed.
+
+In the pull request that removes a package, we will also ping the people who have modified or updated the package, excluding those who only touched the package as part of a tree-wide update.
+
+# Examples and Interactions
+
+[examples-and-interactions]: #examples-and-interactions
+
+- A release manager or another contributor marks `packageA` as broken in 23.11, notifying the maintainer.
+- They have one full release cycle, as described, to fix `packageA`.
+- After the 24.05 release, `packageA` can be automatically removed from the nixpkgs master branch, as outlined in this RFC.
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+Some unmaintained packages may still have users who will no longer be able to use them, but the hope is that removing unmaintained packages will encourage these users to step up as maintainers.
+In such cases, recovering the package from the git history should be relatively easy.
+
+# Alternatives
+
+[alternatives]: #alternatives
+
+- Keep all packages.
+- Archive removed nix expressions. Note that this RFC does not prevent this effort from being implemented in addition.
+
+# Prior Art
+
+[prior-art]: #prior-art
+
+- Most Linux distributions have rules for removing unmaintained packages, e.g., [Debian](https://www.debian.org/doc/manuals/developers-reference/developers-reference.en.html#orphaning-a-package).
+- Policies in [Gentoo](https://wiki.gentoo.org/wiki/Project:Treecleaner), also see [this](https://wiki.gentoo.org/wiki/Project:Treecleaner/Policy)
+  and [this](https://devmanual.gentoo.org/ebuild-maintenance/removal/index.html).
+- [Nixpkgs eval time is increasing too fast](https://github.com/NixOS/nixpkgs/issues/320528): This issue discusses ways to improve evaluation time.
+
+# Unresolved Questions
+
+[unresolved]: #unresolved-questions
+
+None.
+
+# Future Work
+
+[future]: #future-work
+
+- Develop automation that opens pull requests to remove packages.
+- Add deprecation warnings for packages scheduled for removal to notify potential users.
```

Comment:
This should make use of the package problems RFC

```Suggestion
- Add deprecation warnings for packages scheduled for removal to notify potential users using [RFC0127 problems](https://github.com/NixOS/rfcs/blob/master/rfcs/0127-issues-warnings.md)
```

Reply:
I don't see rfc0127 implemented anywhere. We can amend it to the RFC, if it ever get implemented.

Reply:
There’s an open PR (https://github.com/NixOS/nixpkgs/pull/338267). Since this is for future work it seems okay to mention. I thought RFC text was supposed to be immutable after approval?

Reply:
I added it in as an example, but I leave it open if someone else finds a different way instead.